### ISSUE TITLE
Send discord emojis in text from matrix to discord

### ISSUE THREAD

**Issue by XHawk87** [mautrix/discord#82]
I have this bridge on my matrix server, and I have it set up to puppet my discord account. I was trying out direct messages with a friend to see how it works. It's great in most respects, but one key thing that's missing is the ability to use my own discord emojis in chat messages. I wanted to open up a discussion about if/how this can be implemented. I'm a software developer, and I'd be willing to pitch in to help make this happen.

**Comment by XHawk87** [comment]
It seems to me like the correct way to handle this would be to detect `<img data-mx-emoticon/>` tags in `org.matrix.custom.html` messages, and have the discord puppet replace these with the matching discord emoji if found, or just send the image link if not. Any thoughts on using this approach?

**Comment by XHawk87** [comment]
According to https://discord.com/developers/docs/reference#message-formatting

| Type | Structure | Example |
|--|--|--|
| Custom Emoji | `<:NAME:ID>` |  `<:mmLol:216154654256398347>` |
| Custom Emoji (Animated) | `<a:NAME:ID>` | `<a:b1nzy:392938283556143104>` |

So we'll need to get the emoji ID somehow.

**Comment by Spaenny** [comment]
You could retrieve all emotes from a Discord guild using the Discord API (https://discord.com/developers/docs/resources/emoji), and then upload them to the respective space using the State Event: im.ponies.room_emotes. This feature is supported by certain clients such as Cinny and FluffyChat.

**Comment by Spacefreak18** [comment]
for what it's worth, if your client supports text reactions, and you know the name and id of the emote on discord, you can send the emote as a reaction from matrix with

`
/react :emotename:123456789
`

**Comment by Spacefreak18** [comment]
> 

do you atleast see the emotes on the matrix side if they've been added to a message on the discord side?

That's atleast my experience with nheko and schildi chat

**Comment by XHawk87** [comment]
> do you atleast see the emotes on the matrix side if they've been added to a message on the discord side?

Yes, that was my experience too. It's going the other way that was the problem.

Unfortunately, I am not in a position to contribute to this project at the moment. I haven't used this service in a while, it actually stopped working on my homeserver entirely at one point, and I didn't have time to spend troubleshooting it so I disabled it. It seems like a neat idea though, I might come back to it at some point in the future.

**Comment by Spacefreak18** [comment]
> > do you atleast see the emotes on the matrix side if they've been added to a message on the discord side?
> 
> Yes, that was my experience too. It's going the other way that was the problem.
> 
> Unfortunately, I am not in a position to contribute to this project at the moment. I haven't used this service in a while, it actually stopped working on my homeserver entirely at one point, and I didn't have time to spend troubleshooting it so I disabled it. It seems like a neat idea though, I might come back to it at some point in the future.

interesting, because i just recently moved to this because mx-puppet stopped working.

things seem to be stable right now with the newest synapse

i may try to add this functionality myself

**Comment by faithCD** [comment]
First off, thanks to @Spacefreak18 for making this possible for me. I've wrote down some [spaghetti python code](https://github.com/faithCD/mautrix-discord-emoji-pipe) to bridge the discord emojis to matrix while also generating the .sql file for insertion, and also figured out [a hacky workaround](https://github.com/faithCD/mautrix-discord) to make mautrix-discord ignore the `<img data-mx-emoticon>` element during its parsing of a matrix message formatted body. I hope this helps anyone that wishes to bridge emotes in their personal homeservers until mautrix/discord does a proper general-case implementation of bridging custom emotes.
